src/main/java/org/jenkinsci/plugins/workflow/actions/ArgumentsAction.java

```diff
@@ -24,65 +24,137 @@
 
 package org.jenkinsci.plugins.workflow.actions;
 
+import com.google.common.primitives.Primitives;
+import org.apache.commons.collections.CollectionUtils;
 import org.jenkinsci.plugins.workflow.graph.FlowNode;
 import org.jenkinsci.plugins.workflow.graph.StepNode;
 import org.jenkinsci.plugins.workflow.steps.Step;
 import org.jenkinsci.plugins.workflow.steps.StepDescriptor;
 
 import javax.annotation.CheckForNull;
 import javax.annotation.Nonnull;
+import java.util.Collection;
 import java.util.Collections;
 import java.util.HashMap;
+import java.util.List;
 import java.util.Map;
+import java.util.Set;
 
 /**
  * Stores some or all of the arguments used to create and configure the {@link Step} executed by a {@link FlowNode}.
  * This allows you to inspect information supplied in the pipeline script and otherwise discarded at runtime.
  * Supplied argument values can be hidden and replaced with a {@link NotStoredReason} for security or performance.
+ *
+ * Important note: these APIs do not provide recursive guarantees that returned datastructures are immutable.
  */
 public abstract class ArgumentsAction implements PersistentAction {
 
-    /** Used as a placeholder for arguments not stored for various reasons */
+    /** Used as a placeholder marker for {@link Step} arguments not stored for various reasons. */
     public enum NotStoredReason {
-        /** Denotes an unsafe value that cannot be stored/displayed due to sensitive info */
+        /** Denotes an unsafe value that cannot be stored/displayed due to sensitive info. */
         MASKED_VALUE,
 
-        /** Denotes an object that is too big to retain, such as strings exceeding {@link #MAX_STRING_LENGTH} */
+        /** Denotes an object that is too big to retain, such as strings exceeding {@link #MAX_RETAINED_LENGTH} */
         OVERSIZE_VALUE
     }
 
-    /** Largest string we'll persist in the Step, for performance reasons */
-    public static final int MAX_STRING_LENGTH = 1024;
+    /** Largest String, Collection, or array size we'll retain -- provides a rough size limit on any single field.
+     *  Set to 0 or -1 to remove length limits.
+     */
+    private static final int MAX_RETAINED_LENGTH = Integer.getInteger(ArgumentsAction.class.getName()+".maxRetainedLength", 1024);;
+
+    /**
+     * Provides a basic check if an object contains any excessively large single elements.
+     * This recursively checks for String length, Map size, and array/collection size.
+     * We're not trying to cap total size just identify problem elements.
+     * @param o Object to check, with null allowed since we may see null inputs
+     * @param maxElements Max number of elements for a collection/map or characters in a string, or <0 to ignore length rules.
+     * @return True if object (or one of the contained objects) exceeds maxElements size.
+     */
+    public static boolean isOverSized(@CheckForNull Object o, final int maxElements) {
```

Comment:
AFAICT this is something that is only going to be called by the implementation (currently in `workflow-cps`; sooner or later will need to be moved into `workflow-support`), so it would better be moved downstream—there is no reason for it to be in `workflow-api`.